Community-Supported/clouddb-extractor/redshift_extractor.py

```diff
@@ -0,0 +1,142 @@
+    - tableau_username (string): Tableau username
+    - tableau_password (string): Tableau password
+    NOTE: Authentication to Tableau Server can be either by Personal Access Token or
+     Username and Password.  If both are specified then token takes precedence.
+    """
+
+    def __init__(
+        self,
+        source_database_config: dict,
+        tableau_hostname: str,
+        tableau_project: str,
+        tableau_site_id: str = DEFAULT_SITE_ID,
+        tableau_token_name: Optional[str] = None,
+        tableau_token_secret: Optional[str] = None,
+        tableau_username: Optional[str] = None,
+        tableau_password: Optional[str] = None,
+    ) -> None:
+        super().__init__(
+            source_database_config=source_database_config,
+            tableau_hostname=tableau_hostname,
+            tableau_project=tableau_project,
+            tableau_site_id=tableau_site_id,
+            tableau_token_name=tableau_token_name,
+            tableau_token_secret=tableau_token_secret,
+            tableau_username=tableau_username,
+            tableau_password=tableau_password,
+        )
+        self._source_database_connection = None
+        self.sql_identifier_quote = ""
+
+    def source_database_cursor(self) -> Any:
+        """
+        Returns a DBAPI Cursor to the source database
+        """
+        if self._source_database_connection is None:
+            db_connection_args = self.source_database_config.get("connection")
+            self._source_database_connection = redshift_connector.connect(**db_connection_args)
+
+        return self._source_database_connection.cursor()
+
+    def hyper_sql_type(self, source_column: Any) -> SqlType:
+        """
+        Finds the corresponding Hyper column type for source_column
+
+        source_column (obj): Instance of DBAPI Column description tuple
+
+        Returns a tableauhyperapi.SqlType Object
+        """
+
+        type_lookup = {
+            16: SqlType.bool(),
+            17: SqlType.bytes(),
+            1082: SqlType.date(),
+            1114: SqlType.timestamp(),
+            20: SqlType.big_int(),
+            21: SqlType.int(),
+            1700: SqlType.numeric(18, 9),
+            701: SqlType.double(),
+            1043: SqlType.text(),
+            1182: SqlType.time(),
+            1184: SqlType.timestamp_tz(),
+        }
+        source_column_type = source_column[1]
+        return_sql_type = type_lookup.get(source_column_type)
+        if return_sql_type is None:
+            error_message = "No Hyper SqlType defined for Redshift source type: {}".format(source_column_type)
+            logger.error(error_message)
+            raise HyperSQLTypeMappingError(error_message)
+
+        logger.debug("Translated source column type {} to Hyper SqlType {}".format(source_column_type, return_sql_type))
+        return return_sql_type
+
+    def hyper_table_definition(self, source_table: Any, hyper_table_name: str = "Extract") -> TableDefinition:
```

Comment:
this function seems to be identical across all the extractors. Can we move it to the base class, instead of implementing it multiple times?

Reply:
I tried but there really is very little standardization in Cursor.description implementation (e.g. some emit objects, some lists, Redshift needs you to pass column name through decode())